Order List redesign - Step 2: Enhance search support to the order list UI

anitaa1990 opened the pull request:
Fixes #1400. This PR completes the second step of the Stats Improvements Project as defined in the master thread [here](https://github.com/woocommerce/woocommerce-android/issues/1398). 

### Changes
- Added a new class `OrderStatusListView` that displays the order status list UI when search icon is selected.
- Removed the filter option from the toolbar menu.

**Note that this PR is to be merged to a feature branch.**

### Screenshots
<img width="300" src="https://user-images.githubusercontent.com/22608780/65567380-abea4180-df73-11e9-8fe4-c1b4b9273a06.gif"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/65567382-adb40500-df73-11e9-9413-3f5280e7be8c.gif"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/65567417-dc31e000-df73-11e9-9ca7-db85f79a68d6.png">


### Testing
- Search for an order by clicking on an order status.
- Search for an order by typing a search query.
- Verify that pagination, search, filter works as expected.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

nbradbury commented:
If you select a filter and then tap to go back to the full order list, until the list refreshes you continue to see the filtered orders. In the screenshot below I just returned from viewing pending orders.

![Screenshot_1569522881](https://user-images.githubusercontent.com/3903757/65715364-f3072e00-e06a-11e9-8321-3a0cef4db9fe.png)

nbradbury commented:
If you tap the options menu after selecting a filter, the filter is lost.

![untitled](https://user-images.githubusercontent.com/3903757/65715511-3cf01400-e06b-11e9-8a3d-91f3f7d920d0.gif)

nbradbury commented:
After choosing a filter, once you scroll to fetch more posts the filter is lost. Here I'm viewing "on hold" orders, but other orders are in the list.

![Screenshot_1569523222](https://user-images.githubusercontent.com/3903757/65715889-f9e27080-e06b-11e9-9041-ef2f22f8c800.png)

nbradbury commented:
When you select a filter, it briefly shows the "No matching orders" message until it's replaced by the skeleton.

nbradbury commented:
The filter is also lost when returning from order detail.

![untitled](https://user-images.githubusercontent.com/3903757/65716546-60b45980-e06d-11e9-9f00-6b209a658539.gif)

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderListFragment.kt:
```diff
@@ -309,15 +282,19 @@ class OrderListFragment : TopLevelFragment(), OrderListContract.View,
                 presenter.searchOrders(searchQuery)
             } else {
+                // update the toolbar elevation if child fragment is displayed
+                activity?.toolbar?.elevation = if (isChildFragmentShowing()) {
```
I noticed we're changing the toolbar elevation in several places. Would it make sense to add a function to do this?

```
private fun enableToolbarElevation(enable: Boolean) {
    activity?.toolbar?.elevation = if (enable) resources.getDimension(R.dimen.appbar_elevation) else 0f
}
```

anitaa1990 replied:
Good point! Fixed in f1a4030

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderListFragment.kt:
```diff
@@ -617,11 +592,75 @@ class OrderListFragment : TopLevelFragment(), OrderListContract.View,
         searchMenuItem?.setOnActionExpandListener(null)
         searchView?.setOnQueryTextListener(null)
+        hideOrderStatusListView()
+        tab_layout.visibility = View.VISIBLE
```
I find it a little jarring how the tabs suddenly reappear. Maybe a simple animation would help?

```
private fun showTabs(show: Boolean) {
    if (show && tab_layout.visibility != View.VISIBLE) {
        WooAnimUtils.fadeIn(tab_layout)
    } else if (!show && tab_layout.visibility != View.GONE) {
        tab_layout.visibility = View.GONE
    }
}
```

nbradbury commented:
This design doesn't work very well in landscape, but I'm not sure how to address this.

![Screenshot_1569525060](https://user-images.githubusercontent.com/3903757/65717989-99096700-e070-11e9-85ef-2544654b31e9.png)

AmandaRiu commented:
That's weird...I had assigned this to myself and spent a lot of time reviewing it but came back in and after posting my review noticed it had been unassigned from me?

anitaa1990 commented on RELEASE-NOTES.txt:
```diff
@@ -1,6 +1,7 @@
 2.8
 -----
 * Revamped the order list page by providing a new tab for unfulfilled orders
+* Added new interface to view order statuses with their count and provided option to search orders by order status.
```
👌 Fixed in dc1f0f9

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderListFragment.kt:
```diff
@@ -617,11 +592,75 @@ class OrderListFragment : TopLevelFragment(), OrderListContract.View,
+            val tabPosition = AppPrefs.getSelectedOrderListTabPosition()
+            orderStatusFilter = tab_layout.getTabAt(tabPosition)?.let { getOrderStatusByTab(it) }
+
+            presenter.loadOrders(orderStatusFilter, forceRefresh = isFilterEnabled)
```
Fixed in f616410

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderListFragment.kt:
```diff
@@ -617,11 +592,75 @@ class OrderListFragment : TopLevelFragment(), OrderListContract.View,
+    }
+
+    /**
+     * Method called when user clicks on an order status from [OrderStatusListView]
```
👍 Fixed in f616410

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderListFragment.kt:
```diff
@@ -617,11 +592,75 @@ class OrderListFragment : TopLevelFragment(), OrderListContract.View,
+    }
+
+    /**
+     * Method called when user clicks on the back button after selecting an order status
```
Fixed in f616410

anitaa1990 commented on WooCommerce/src/main/res/layout/order_status_list_item.xml:
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Done in f616410

anitaa1990 commented:
@nbradbury @AmandaRiu thank you for the thorough review!

> If you tap the options menu after selecting a filter, the filter is lost.

Good catch! This was because when another menu item is clicked, the search is automatically collapsed. There wasn't a straightforward fix for this so in b979ca1 added logic to hide the menu option when search is enabled.

> After choosing a filter, once you scroll to fetch more posts the filter is lost.

I wasn't able to reproduce this behaviour but I did notice that no orders were loaded on pagination when filtered was enabled. Fixed that in 2667302 and I _think_  this might have been related to the issue you reported. Can you test and let me know if you are able to reproduce this now?

> When you select a filter, it briefly shows the "No matching orders" message until it's replaced by the skeleton.

I believe this issue persists in `develop` as well but I added a fix in 67e6717 to hide the empty view if the order list is refreshing.

> If you select a filter and then tap to go back to the full order list, until the list refreshes you continue to see the filtered orders. 

This is the current behaviour from `develop` as well. Even when we click on the `All orders` tab, the Processing orders are displayed while the refresh indicator is loading and only when loaded the order list is refreshed. I think it would make more sense to be consistent with this behaviour and change it everywhere rather than just when clicking back from the filtered list. What do you think? 

> The filter is also lost when returning from order detail.

Fixed in 556d4d5

> This design doesn't work very well in landscape, but I'm not sure how to address this.

I'm not sure if this is the ideal experience but I've added logic to close the keypad when in landscape and only if the user clicks on the search will they be able to search by query. This highlights the order status list by default. 

> - The search hint in the design is "Search orders", but in the implementation it's "Search..."
- The color of the status count seems to be off... should be the the same as the "order status" list label

Fixed in d5e4a3f

> The cursor in the search box is the same color as the background of the toolbar so it's essentially invisible and it makes it difficult to know the search text view is active. I think Nick is currently working on a similar fix for Products.

I believe Nick has a [PR for review](https://github.com/woocommerce/woocommerce-android/pull/1427) to address the search cursor in both Products and Orders so leaving it as is for now.

> Clicking on a status when in the search view to filter by that status when the device is offline will just show the loading skeleton indefinitely if nothing downloaded for that status. The search view will also show the loading skeleton indefinitely if offline and order status totals were never downloaded. 
If you are offline and select a filter status in the search screen, the loading skeleton is displayed. Bringing the device back online does not load the data.

Good point! Fixed in db93363. Screenshot the current behaviour below.

<img width="300" src="https://user-images.githubusercontent.com/22608780/65860940-f52dfd00-e388-11e9-9a5e-5a320b597c65.gif">

The same behaviour is replicated for order status list as well. I've added an option to refresh the order status list when internet becomes available. At the moment though, if there are no order status options stored locally and if no internet is available, the screen is blank. I'm not sure how we should handle this scenario. Should we display an empty message stating: "Order statuses not available" or should we just display the search view as it is now i.e. A blank search screen waiting for user's input?

> The order status totals in the search view are not accurate. See the screenshots below. They also don't seem to update after an order status change. 

The order status count is not updating after an order status change. This is a bug I found when working on handling empty states for order list issue [here](https://github.com/woocommerce/woocommerce-android/issues/1411) and the fix has been added to FluxC in this [PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1389). This updated fluxC hash has been added to the order list redesign - step 3 [PR](https://github.com/woocommerce/woocommerce-android/pull/1430). 

I'm not sure why there is a difference between the web count and the local count though. For the same test site, on my device, the count is displayed correctly 🤔.

<img width="1394" src="https://user-images.githubusercontent.com/22608780/65861699-573b3200-e38a-11e9-8847-c1cf07a1bfd3.png">

> The extra layer of tabs make viewing the orders list in landscape mode really squished feeling. I recommend implementing scrolling so it collapses the main toolbar like the way WPAndroid does with the notifications tab to give the user more room to view the orders list

I'm not sure how to go about with this change. I looked at the WPAndroid notifications tab and they use an activity layout and have the toolbar and the `TabLayout` inside the same [xml file](https://github.com/wordpress-mobile/WordPress-Android/blob/10a7da05bf2310060f46d61aa50c8bf449ebe5d7/WordPress/src/main/res/layout/comment_activity.xml#L3). We use a common activity layout and the `TabLayout` is added only for the orders fragment. I'm not able to come up with an easy implementation for this and foresee a lot of rework to the `MainActivity` and the Fragment files to achieve this. Since this is the current behaviour for the stats v4 UI as well, it would make more sense to come up with a cleaner implementation for this in a separate PR. WDYT?

> Orders are displayed filtered by the order status when clicking on an order status in the search list, but clicking "back" takes the user back to the order list. 

Fixed in 18a2551 by adding logic to display the order list status screen again when back is clicked. Screenshot of interaction  below:

<img width="300" src="https://user-images.githubusercontent.com/22608780/65862335-a897f100-e38b-11e9-95cd-efbcc865bb20.gif">

@nbradbury @AmandaRiu I believe all the issues have been addressed and PR is ready for another round 👍

nbradbury commented:
> I wasn't able to reproduce this behaviour but I did notice that no orders were loaded on pagination when filtered was enabled. Fixed that in [2667302](https://github.com/woocommerce/woocommerce-android/commit/26673028fb0aee49a2dae50c533d8395cf540e6e) and I _think_ this might have been related to the issue you reported. Can you test and let me know if you are able to reproduce this now?

This change appears to have fixed it :)

nbradbury commented:
@AmandaRiu I'm good with the changes to the issues I raised. Since you were the original reviewer for this PR, I'll defer to you as to when this is merged.

AmandaRiu commented:
Hi Anitaa! Here are the results of my latest review:

## Offline Mode Funkiness

If there are some cached results available when filtering by an order status they will show up for a second and then disappear:

[offline-search-funky]


The "Processing tab" will show "no orders to process", but clicking on the "all orders" tab will show orders with the status of "processing". Switching between the "processing" and "all orders" tab in this scenario will also momentarily display the "Share your store" image.

[offline-processing-tab-empty]



If you are offline on the search view and then go online, the search view will immediately populate. But if you got back online and then cancel search the processing tab will show an indefinite skeleton until you switching between the "processing" and "all orders" tab.

[offline-online-processing]

## Other Stuff

> I wasn't able to reproduce this behaviour but I did notice that no orders were loaded on pagination when filtered was enabled. Fixed that in 2667302 and I think this might have been related to the issue you reported. Can you test and let me know if you are able to reproduce this now?

I was not able to reproduce this so it must be fixed! 

> This is the current behaviour from develop as well. Even when we click on the All orders tab, the Processing orders are displayed while the refresh indicator is loading and only when loaded the order list is refreshed. I think it would make more sense to be consistent with this behaviour and change it everywhere rather than just when clicking back from the filtered list. What do you think?

I agree. We can address this separately - maybe after the new order list management is in place.

> I'm not sure if this is the ideal experience but I've added logic to close the keypad when in landscape and only if the user clicks on the search will they be able to search by query. This highlights the order status list by default.

That works, but I still also think adding the collapsing toolbar would go a log way as well. This is what I had recommended:

> The extra layer of tabs make viewing the orders list in landscape mode really squished feeling. I recommend implementing scrolling so it collapses the main toolbar like the way WPAndroid does with the notifications tab to give the user more room to view the orders list

> I believe Nick has a PR for review to address the search cursor in both Products and Orders so leaving it as is for now.

Sounds good!

> At the moment though, if there are no order status options stored locally and if no internet is available, the screen is blank. I'm not sure how we should handle this scenario. Should we display an empty message stating: "Order statuses not available" or should we just display the search view as it is now i.e. A blank search screen waiting for user's input?

I agree, what the user sees now is kinda buggy looking so I'm wondering if maybe we should have a special "offline" view for this as well? Cc: @kyleaparker - here's a screenshot for reference:

[offline-order-status-empty]

> The order status count is not updating after an order status change. This is a bug I found when working on handling empty states for order list issue here and the fix has been added to FluxC in this PR. This updated fluxC hash has been added to the order list redesign - step 3 PR.

Sounds good!

> I'm not sure why there is a difference between the web count and the local count though. For the same test site, on my device, the count is displayed correctly 🤔.

Try making some major changes in the web client - none of them will be updated in the app. It's the same if you make order status changes inside the app.

> I'm not sure how to go about with this change. I looked at the WPAndroid notifications tab and they use an activity layout and have the toolbar and the TabLayout inside the same xml file. We use a common activity layout and the TabLayout is added only for the orders fragment. I'm not able to come up with an easy implementation for this and foresee a lot of rework to the MainActivity and the Fragment files to achieve this. Since this is the current behaviour for the stats v4 UI as well, it would make more sense to come up with a cleaner implementation for this in a separate PR. WDYT?

That sounds like a good plan for now.

That's it for this round!

anitaa1990 commented:
Thank you for a thorough feedback @AmandaRiu!

>  If there are some cached results available when filtering by an order status they will show up for a second and then disappear:

> The "Processing tab" will show "no orders to process", but clicking on the "all orders" tab will show orders with the status of "processing". Switching between the "processing" and "all orders" tab in this scenario will also momentarily display the "Share your store" image.

> If you are offline on the search view and then go online, the search view will immediately populate. But if you got back online and then cancel search the processing tab will show an indefinite skeleton until you switching between the "processing" and "all orders" tab.

I believe all the offline chunkiness was caused by the same bug. Fixed in 5aceb16

> I agree, what the user sees now is kinda buggy looking so I'm wondering if maybe we should have a special "offline" view for this as well.

I'm not sure if this is the best experience but I added an option to remove the list entirely when there is no internet in a2b5bf7. This is the screenshot of what happens when there is no internet and the order status list is not cached locally:

<img width="300" src="https://user-images.githubusercontent.com/22608780/66058832-ab940c80-e558-11e9-9bc6-8d241aacb452.png">

> 'm not sure why there is a difference between the web count and the local count though. For the same test site, on my device, the count is displayed correctly 🤔.

> Try making some major changes in the web client - none of them will be updated in the app. It's the same if you make order status changes inside the app.

Yup, this looks like the same bug I found and fixed in FluxC in this [PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1389) This updated fluxC hash has been added to the order list redesign - step 3 [PR](https://github.com/woocommerce/woocommerce-android/pull/1430).

PR ready for another round 👍

kyleaparker commented:
>I agree, what the user sees now is kinda buggy looking so I'm wondering if maybe we should have a special "offline" view for this as well? Cc: @kyleaparker - here's a screenshot for reference:

I'm not sure we need another message there since we are already showing an offline banner, I don't think the additional message is adding anything. If anything, it would be best to rewrite the offline message to better communicate what's happening. For example: "Offline - Some data can't be loaded"

It looks a bit odd that we show the bottom navigation in the search view, can we remove this on search views?

jkmassel commented:
Howdy! We're freezing `2.8` today, so I'm bumping this PR to `2.9`. If it needs to be released as part of `2.8`, please target it to the `release/2.8` branch, and DM me once it's merged. I'll be happy to do a new beta build!

anitaa1990 commented:
Thanks for the thorough review @AmandaRiu.

> The only thing I have noticed is that the "Processing" tab doesn't always populate when going from offline to online. 

Fixed in  002809f

As @kyleaparker pointed out, I also removed the bottom nav when search is enabled in f32d704. Ready for another round 👍
